File: packages/domain/src/exclusion.ts
import { GameCode, type HistoricalDraw } from "@lottery/contracts";

export interface Candidate {
  gameCode: GameCode;
  primaryNumbers: readonly number[];
  secondaryNumbers: readonly number[];
}

export function asHistoricalDraws(
  draws: Array<{
    issue: string;
    primaryNumbers: number[];
    secondaryNumbers: number[];
  }>,
): HistoricalDraw[] {
  return draws.map((draw) => ({
    issue: draw.issue,
    primaryNumbers: new Set(draw.primaryNumbers),
    secondaryNumbers: new Set(draw.secondaryNumbers),
  }));
}

export function wouldMatchHistoricalHighPrize(
  candidate: Candidate,
  historicalDraws: HistoricalDraw[],
): { excluded: boolean; issue: string | null } {
  const primary = new Set(candidate.primaryNumbers);
  const secondary = new Set(candidate.secondaryNumbers);

  for (const draw of historicalDraws) {
    if (candidate.gameCode === GameCode.DOUBLE_COLOR_BALL) {
      if (isSubset(draw.primaryNumbers, primary)) {
        return { excluded: true, issue: draw.issue };
      }
      continue;
    }
    if (isSubset(draw.primaryNumbers, primary) && intersects(draw.secondaryNumbers, secondary)) {
      return { excluded: true, issue: draw.issue };
    }
  }
  return { excluded: false, issue: null };
}

function isSubset(source: ReadonlySet<number>, target: ReadonlySet<number>): boolean {
  for (const value of source) {
    if (!target.has(value)) {
      return false;
    }
  }
  return true;
}

function intersects(left: ReadonlySet<number>, right: ReadonlySet<number>): boolean {
  for (const value of left) {
    if (right.has(value)) {
      return true;
    }
  }
  return false;
}

File: server/app/domain/exclusion.py
from dataclasses import dataclass

from app.domain.games import GameCode


@dataclass(frozen=True)
class HistoricalDraw:
    issue: str
    primary_numbers: frozenset[int]
    secondary_numbers: frozenset[int]


@dataclass(frozen=True)
class Candidate:
    game_code: GameCode
    primary_numbers: tuple[int, ...]
    secondary_numbers: tuple[int, ...]

    @property
    def primary_set(self) -> frozenset[int]:
        return frozenset(self.primary_numbers)

    @property
    def secondary_set(self) -> frozenset[int]:
        return frozenset(self.secondary_numbers)


def would_match_historical_high_prize(
    candidate: Candidate,
    historical_draws: list[HistoricalDraw],
) -> tuple[bool, str | None]:
    primary = candidate.primary_set
    secondary = candidate.secondary_set

    for draw in historical_draws:
        if candidate.game_code == GameCode.DOUBLE_COLOR_BALL:
            if draw.primary_numbers.issubset(primary):
                return True, draw.issue
        elif candidate.game_code == GameCode.SUPER_LOTTO:
            if draw.primary_numbers.issubset(primary) and draw.secondary_numbers & secondary:
                return True, draw.issue
    return False, None

